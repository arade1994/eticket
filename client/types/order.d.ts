import { type Ticket } from "./ticket";

export interface Order {
  id: string;
  userId: string;
  status: OrderStatus;
  expiresAt: Date;
  version: number;
  ticket: Ticket;
}

export declare enum OrderStatus {
  Created = "created",
  Cancelled = "cancelled",
  AwaitingPayment = "awaiting:payment",
  Complete = "complete",
}
