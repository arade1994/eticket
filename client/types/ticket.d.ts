export interface Ticket {
  id: string;
  title: string;
  price: number;
  userId: string;
  category: string;
  version: number;
}
