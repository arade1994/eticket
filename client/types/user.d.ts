import { type IncomingMessage } from "http";

export interface User {
  id: string;
  firstName: string;
  lastName: string;
  age: number;
  email: string;
  password: string;
}

export interface Rating {
  id: string;
  comment: string;
  rate: number;
  userId: string;
  ratedUserId: string;
}

export interface RequestWithUser extends IncomingMessage {
  currentUser?: { id: string; email: string };
}
