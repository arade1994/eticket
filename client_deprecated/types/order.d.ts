import { type Ticket } from "./ticket";

export interface Order {
  id: string;
  userId: string;
  status: OrderStatus;
  expiresAt: string;
  version: number;
  ticket: Ticket;
}

export declare enum OrderStatus {
  Created = "created",
  Cancelled = "cancelled",
  AwaitingPayment = "awaiting:payment",
  Complete = "complete",
}
