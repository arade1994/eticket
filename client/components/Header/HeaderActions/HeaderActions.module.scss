.headerActions {
  display: flex;
  align-items: center;
  gap: 1rem;

  & .notifications {
    position: relative;

    & .notificationButton {
      background: rgba(255, 255, 255, 0.15);
      border: 2px solid rgba(255, 255, 255, 0.3);
      padding: 0.5rem;
      cursor: pointer;
      position: relative;
      color: #ffffff;
      border-radius: 50%;
      width: 40px;
      height: 40px;
      display: flex;
      align-items: center;
      justify-content: center;
      transition: all 0.3s ease;
      text-shadow: 0 1px 2px rgba(0, 0, 0, 0.1);

      & .notificationIcon {
        font-size: 1.25rem;
        opacity: 0.95;
        transition: all 0.3s ease;
      }

      & .notificationBadge {
        position: absolute;
        top: -2px;
        right: -2px;
        width: 10px;
        height: 10px;
        background-color: #ff4757;
        border-radius: 50%;
        border: 2px solid #006064;
        box-shadow: 0 1px 2px rgba(0, 0, 0, 0.2);
        animation: pulse 2s infinite;
      }

      &:hover {
        background: rgba(255, 255, 255, 0.25);
        border-color: rgba(255, 255, 255, 0.4);
        transform: translateY(-1px);
        box-shadow: 0 2px 4px rgba(0, 0, 0, 0.1);
      }

      &:active {
        transform: translateY(0);
      }
    }

    & .notificationDropdown {
      position: absolute;
      top: 100%;
      right: 0;
      width: 320px;
      background-color: var(--color-white);
      border-radius: var(--border-radius);
      box-shadow: var(--shadow-lg);
      margin-top: 0.5rem;
      display: none;

      & .notificationHeader {
        padding: 1rem;
        border-bottom: 1px solid var(--color-border);
        display: flex;
        justify-content: space-between;
        align-items: center;

        h3 {
          margin: 0;
          font-size: 1rem;
          font-weight: 600;
        }

        & .clearButton {
          background: none;
          border: none;
          color: var(--color-text-light);
          cursor: pointer;
          font-size: 0.875rem;
          padding: 0.25rem 0.5rem;
          border-radius: var(--border-radius);
          transition: all 0.3s ease;

          &:hover {
            background-color: var(--color-background);
            color: var(--color-text);
          }
        }
      }

      & .notificationList {
        max-height: 400px;
        overflow-y: auto;

        & .notificationItem {
          padding: 1rem;
          border-bottom: 1px solid var(--color-border);
          cursor: pointer;
          transition: background-color 0.3s ease;

          &:hover {
            background-color: var(--color-background);
          }

          & .unread {
            background-color: var(--color-primary-light);
          }

          p {
            margin: 0;
            font-size: 0.875rem;
            color: var(--color-text);
          }
        }

        & .notificationTime {
          font-size: 0.75rem;
          color: var(--color-text-light);
          margin-top: 0.25rem;
          display: block;
        }
      }

      & .open {
        display: block;
      }
    }
  }

  & .signOutButton {
    background: rgba(255, 255, 255, 0.15);
    border: 2px solid rgba(255, 255, 255, 0.3);
    color: #ffffff;
    padding: 0.5rem 1.25rem;
    font-weight: 600;
    letter-spacing: 0.3px;
    text-shadow: 0 1px 2px rgba(0, 0, 0, 0.1);
    transition: all 0.3s ease;

    &:hover {
      background: rgba(255, 255, 255, 0.25);
      border-color: rgba(255, 255, 255, 0.4);
      transform: translateY(-1px);
      box-shadow: 0 2px 4px rgba(0, 0, 0, 0.1);
    }

    &:active {
      transform: translateY(0);
    }
  }
}

.signInButton,
.signOutButton {
  padding: 0.5rem 1rem;
  border-radius: var(--border-radius);
  font-weight: 500;
  transition: all 0.3s ease;
  cursor: pointer;
}

.signInButton {
  background: rgba(255, 255, 255, 0.15);
  border: 2px solid rgba(255, 255, 255, 0.3);
  color: #ffffff;
  padding: 0.5rem 1.25rem;
  font-weight: 600;
  letter-spacing: 0.3px;
  text-shadow: 0 1px 2px rgba(0, 0, 0, 0.1);
  transition: all 0.3s ease;
  text-decoration: none;
  border-radius: var(--border-radius);

  &:hover {
    background: rgba(255, 255, 255, 0.25);
    border-color: rgba(255, 255, 255, 0.4);
    transform: translateY(-1px);
    box-shadow: 0 2px 4px rgba(0, 0, 0, 0.1);
  }

  &:active {
    transform: translateY(0);
  }
}

@keyframes pulse {
  0% {
    transform: scale(1);
    opacity: 1;
  }
  50% {
    transform: scale(1.2);
    opacity: 0.8;
  }
  100% {
    transform: scale(1);
    opacity: 1;
  }
}
