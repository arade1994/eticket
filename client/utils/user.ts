import { Ticket } from "../types/ticket";
import { User } from "../types/user";

export const getNumOfCreatedTickets = (user: User, tickets: Ticket[]) =>
  tickets?.filter((ticket) => ticket.userId === user.id)?.length;
