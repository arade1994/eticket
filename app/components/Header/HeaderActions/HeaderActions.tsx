import Link from "next/link";

import { useAuth } from "../../../hooks/useAuth";
import { type User } from "../../../types/user";

import classes from "./HeaderActions.module.scss";

interface Props {
  currentUser?: User;
}

const HeaderActions: React.FC<React.PropsWithChildren<Props>> = ({
  currentUser,
}) => {
  const { logOut } = useAuth();

  return (
    <div className={classes.headerActions}>
      {currentUser ? (
        <button className={classes.signOutButton} onClick={logOut}>
          Log Out
        </button>
      ) : (
        <>
          <Link className={classes.signInButton} href="/auth/signin">
            Login
          </Link>
          <Link className={classes.signInButton} href="/auth/signup">
            Register
          </Link>
        </>
      )}
    </div>
  );
};

export default HeaderActions;
