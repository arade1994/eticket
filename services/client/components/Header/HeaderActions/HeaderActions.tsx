import Link from "next/link";
import { FaBell } from "react-icons/fa";

import { useAuth } from "../../../hooks/useAuth";
import { useNotification } from "../../../hooks/useNotification";
import { type User } from "../../../types/user";

import classes from "./HeaderActions.module.scss";

interface Props {
  currentUser?: User;
}

const HeaderActions: React.FC<React.PropsWithChildren<Props>> = ({
  currentUser,
}) => {
  const { signOut } = useAuth();
  const { notifications, markAsRead, clearAll } = useNotification();

  return (
    <div className={classes.headerActions}>
      {currentUser ? (
        <>
          <div className={classes.notifications}>
            <button
              className={classes.notificationButton}
              onClick={() => {
                // Toggle notifications dropdown
              }}
            >
              <FaBell aria-label="notifications-bell" />
              {notifications.some((n) => !n.read) && (
                <span className={classes.notificationBadge} />
              )}
            </button>
            <div className={classes.notificationDropdown}>
              <div className={classes.notificationHeader}>
                <h3>Notifications</h3>
                <button className={classes.clearButton} onClick={clearAll}>
                  Clear All
                </button>
              </div>
              <div className={classes.notificationList}>
                {notifications.map((notification) => (
                  <div
                    key={notification.id}
                    className={`${classes.notificationItem} ${
                      !notification.read ? classes.unread : ""
                    }`}
                    onClick={() => markAsRead(notification.id)}
                  >
                    <p>{notification.message}</p>
                    <span className={classes.notificationTime}>
                      {new Date(notification.createdAt).toLocaleDateString()}
                    </span>
                  </div>
                ))}
              </div>
            </div>
          </div>
          <button className={classes.signOutButton} onClick={signOut}>
            Sign Out
          </button>
        </>
      ) : (
        <Link className={classes.signInButton} href="/auth/signin">
          Sign In
        </Link>
      )}
    </div>
  );
};

export default HeaderActions;
