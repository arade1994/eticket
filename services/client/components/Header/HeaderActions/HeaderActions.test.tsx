import { describe, expect, test } from "vitest";

import { render, screen } from "@testing-library/react";

import mockDb from "../../../mock/db.json";
import { type User } from "../../../types/user";

import HeaderActions from "./HeaderActions";

const mockCurrentUser = mockDb.currentUser as unknown as User;

describe("<HeaderActions />", () => {
  test("it should match a snapshot when user is signed in by default", () => {
    const { baseElement } = render(
      <HeaderActions currentUser={mockCurrentUser} />
    );

    expect(baseElement).toMatchSnapshot("Signed In");

    const notificationsBtn = screen.getByLabelText("notifications-bell");
    expect(notificationsBtn).toBeDefined();
    const signOutLink = screen.getByText("Sign Out");
    expect(signOutLink).toBeDefined();
  });

  test("it should match a snapshot when user is signed out by default", () => {
    const { baseElement } = render(<HeaderActions />);

    expect(baseElement).toMatchSnapshot("Signed Out");

    const signInLink = screen.getByText("Sign In");
    expect(signInLink).toBeDefined();
  });
});
