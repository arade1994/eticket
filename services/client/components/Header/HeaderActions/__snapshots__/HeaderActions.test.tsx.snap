// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<HeaderActions /> > it should match a snapshot when user is signed in by default > Signed In 1`] = `
<body>
  <div>
    <div
      class="_headerActions_99caa7"
    >
      <div
        class="_notifications_99caa7"
      >
        <button
          class="_notificationButton_99caa7"
        >
          <svg
            aria-label="notifications-bell"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M224 512c35.32 0 63.97-28.65 63.97-64H160.03c0 35.35 28.65 64 63.97 64zm215.39-149.71c-19.32-20.76-55.47-51.99-55.47-154.29 0-77.7-54.48-139.9-127.94-155.16V32c0-17.67-14.32-32-31.98-32s-31.98 14.33-31.98 32v20.84C118.56 68.1 64.08 130.3 64.08 208c0 102.3-36.15 133.53-55.47 154.29-6 6.45-8.66 14.16-8.61 21.71.11 16.4 12.98 32 32.1 32h383.8c19.12 0 32-15.6 32.1-32 .05-7.55-2.61-15.27-8.61-21.71z"
            />
          </svg>
        </button>
        <div
          class="_notificationDropdown_99caa7"
        >
          <div
            class="_notificationHeader_99caa7"
          >
            <h3>
              Notifications
            </h3>
            <button
              class="_clearButton_99caa7"
            >
              Clear All
            </button>
          </div>
          <div
            class="_notificationList_99caa7"
          />
        </div>
      </div>
      <button
        class="_signOutButton_99caa7"
      >
        Sign Out
      </button>
    </div>
  </div>
</body>
`;

exports[`<HeaderActions /> > it should match a snapshot when user is signed out by default > Signed Out 1`] = `
<body>
  <div>
    <div
      class="_headerActions_99caa7"
    >
      <a
        class="_signInButton_99caa7"
        href="/auth/signin"
      >
        Sign In
      </a>
    </div>
  </div>
</body>
`;
